### PR TITLE
[1265] Handle Flexbox Container on FormDescriptionEditors

### PR DESCRIPTION
**Pull request by AxelRICHARD**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [x] Variables have a proper type
- [x] Functions’ arguments have a proper type
- [x] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [x] `useState<STATE_TYPE>(…)`
- [x] All components have a proper typing for their props
- [x] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [x] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

**Review by gcoutable — DISMISSED** [review]

File: backend/sirius-components-collaborative-formdescriptioneditors/src/test/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/AddWidgetEventHandlerTests.java [hunk 1]
```diff
@@ -41,14 +41,14 @@
  */
 public class AddWidgetEventHandlerTests {
     @Test
-    public void testArrangeAll() {
+    public void testAddwidget() {
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply backend/sirius-components-collaborative-formdescriptioneditors/src/test/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/AddWidgetEventHandlerTests.java:44]: Done

File: frontend/src/formdescriptioneditor/WidgetEntry.types.ts [hunk 2]
```diff
@@ -10,17 +10,25 @@
...
+  GQLFormDescriptionEditorFlexboxContainer,
+  GQLFormDescriptionEditorWidget,
+} from './FormDescriptionEditorEventFragment.types';
 
 export interface WidgetEntryProps {
   editingContextId: string;
   representationId: string;
-  formDescriptionEditor: GQLFormDescriptionEditor;
-  widget: GQLFormDescriptionEditorWidget;
+  containerId: string;
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply frontend/src/formdescriptioneditor/WidgetEntry.types.ts:25]: Done

File: frontend/src/icons/FlexboxContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import PropTypes from 'prop-types';
+import React from 'react';
+
+const propTypes = {
+  title: PropTypes.string.isRequired,
+};
+
+export const FlexboxContainer = ({ title, ...props }) => {
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply frontend/src/icons/FlexboxContainer.tsx:20]: Done

File: frontend/src/properties/Page.tsx [hunk 4]
```diff
@@ -22,6 +22,7 @@ const usePageStyles = makeStyles((theme) => ({
     '& > *': {
       marginBottom: theme.spacing(2),
     },
+    overflowY: 'scroll',
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply frontend/src/properties/Page.tsx:25]: Done

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/AddWidgetInput.java [hunk 5]
```diff
@@ -39,10 +41,13 @@ public AddWidgetInput() {
         // Used by jackson
     }
 
-    public AddWidgetInput(UUID id, String editingContextId, String representationId) {
+    public AddWidgetInput(UUID id, String editingContextId, String representationId, String containerId, String kind, int index) {
         this.id = Objects.requireNonNull(id);
         this.editingContextId = Objects.requireNonNull(editingContextId);
         this.representationId = Objects.requireNonNull(representationId);
+        this.containerId = Objects.requireNonNull(containerId);
```
Comment (sbegaudeau) [inline packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/AddWidgetInput.java:48]: containerId is nullable

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/MoveWidgetInput.java [hunk 6]
```diff
@@ -39,10 +41,13 @@ public MoveWidgetInput() {
         // Used by jackson
     }
 
-    public MoveWidgetInput(UUID id, String editingContextId, String representationId) {
+    public MoveWidgetInput(UUID id, String editingContextId, String representationId, String containerId, String widgetId, int index) {
         this.id = Objects.requireNonNull(id);
         this.editingContextId = Objects.requireNonNull(editingContextId);
         this.representationId = Objects.requireNonNull(representationId);
+        this.containerId = Objects.requireNonNull(containerId);
```
Comment (sbegaudeau) [inline packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/MoveWidgetInput.java:48]: containerId is nullable

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/AddWidgetEventHandler.java [hunk 7]
```diff
@@ -94,23 +98,35 @@ public void handle(One<IPayload> payloadSink, Many<ChangeDescription> changeDesc
         changeDescriptionSink.tryEmitNext(changeDescription);
     }
 
-    protected boolean addWidget(IEditingContext editingContext, IFormDescriptionEditorContext formDescriptionEditorContext, String kind, int index) {
-        var optionalSelf = this.objectService.getObject(editingContext, formDescriptionEditorContext.getFormDescriptionEditor().getTargetObjectId());
+    protected boolean addWidget(IEditingContext editingContext, IFormDescriptionEditorContext formDescriptionEditorContext, String containerId, String kind, int index) {
```
Comment (sbegaudeau) [inline packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/AddWidgetEventHandler.java:101]: why `protected`?

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/MoveWidgetEventHandler.java [hunk 8]
```diff
@@ -90,18 +92,36 @@ public void handle(One<IPayload> payloadSink, Many<ChangeDescription> changeDesc
         changeDescriptionSink.tryEmitNext(changeDescription);
     }
 
-    protected boolean moveWidget(IEditingContext editingContext, IFormDescriptionEditorContext formDescriptionEditorContext, String widgetId, int index) {
-        var optionalSelf = this.objectService.getObject(editingContext, formDescriptionEditorContext.getFormDescriptionEditor().getTargetObjectId());
+    protected boolean moveWidget(IEditingContext editingContext, IFormDescriptionEditorContext formDescriptionEditorContext, String containerId, String widgetId, int index) {
```
Comment (sbegaudeau) [inline packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/MoveWidgetEventHandler.java:95]: why `protected`?

File: packages/formdescriptioneditors/backend/sirius-components-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/formdescriptioneditors/components/FormDescriptionEditorWidgetComponent.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ * The component used to render the form description editor widget.
+ *
+ * @author arichard
+ */
+public class FormDescriptionEditorWidgetComponent implements IComponent {
+
+    private final FormDescriptionEditorWidgetComponentProps props;
+
+    public FormDescriptionEditorWidgetComponent(FormDescriptionEditorWidgetComponentProps props) {
+        this.props = props;
```
Comment (sbegaudeau) [inline packages/formdescriptioneditors/backend/sirius-components-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/formdescriptioneditors/components/FormDescriptionEditorWidgetComponent.java:30]: Objects.requireNonNull

File: packages/formdescriptioneditors/frontend/sirius-components-formdescriptioneditors/src/FlexboxContainerWidget.tsx [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+    padding: '0px 0px 0px 4px',
+    width: '100%',
+  },
+  container: {
+    display: 'flex',
+    flexWrap: ({ flexWrap }) => flexWrap,
+    flexDirection: ({ flexDirection }) => flexDirection,
+    '& > *': {
+      marginBottom: theme.spacing(0),
+    },
```
Comment (sbegaudeau) [inline packages/formdescriptioneditors/frontend/sirius-components-formdescriptioneditors/src/FlexboxContainerWidget.tsx:56]: gap would be better but that's a minor issue

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
I'll handle the changes that I'm requesting. I still have to test it locally and I'll ensure that it's merged after that.

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/AddWidgetEventHandler.java (hunk 7 above)
Comment (sbegaudeau) [reply packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/AddWidgetEventHandler.java:101]: So this method was protected in order to write a test without mocking the object service properly. By writing the test properly, it appears that the test could not really work in normal circumstances and the code below could fail with a bug too.

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/MoveWidgetEventHandler.java (hunk 8 above)
Comment (sbegaudeau) [reply packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/MoveWidgetEventHandler.java:95]: I have removed this protected visibility and updated the test to make it look like a proper test. It gave me the opportunity to fix a small bug in the test too.

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**